clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -8007,10 +8007,29 @@ static bool checkSYCLAddIRAttributesMergeability(const AddIRAttrT &NewAttr,
 
 void Sema::CheckSYCLAddIRAttributesFunctionAttrConflicts(Decl *D) {
   const auto *AddIRFuncAttr = D->getAttr<SYCLAddIRAttributesFunctionAttr>();
-  if (!AddIRFuncAttr || AddIRFuncAttr->args_size() == 0 ||
+
+  // If there is no such attribute there is nothing to check. If there are
+  // dependent arguments we cannot know the actual number of arguments so we
+  // defer the check.
+  if (!AddIRFuncAttr ||
       hasDependentExpr(AddIRFuncAttr->args_begin(), AddIRFuncAttr->args_size()))
     return;
 
+  // If there are no name-value pairs in the attribute it will not have an
+  // effect and we can skip the check. The filter is ignored.
+  size_t NumArgsWithoutFilter =
+      AddIRFuncAttr->args_size() - (AddIRFuncAttr->hasFilterList() ? 1 : 0);
+  if (NumArgsWithoutFilter == 0)
+    return;
+
+  // "sycl-single-task" is present on all single_task invocations, implicitly
+  // added by the SYCL headers. It can only conflict with max_global_work_dim,
+  // but the value will be the same so there is no need for a warning.
+  if (NumArgsWithoutFilter == 2 &&
+      AddIRFuncAttr->getAttributeNameValuePairs(Context)[0].first ==
```

Comment:
I approved the patch already but thinking about it, I had more questions about this logic. The filer list basically specifies which name-value pairs are valid arguments right? The ones not in this list are technically not used correct? If that is the case, isn't it against that list you are supposed to check? Does `getAttributeNameValuePairs` return only those in the filter list?

Also if there is a filter list and `sycl-single-task` is specified in `add_ir_attributes_function` but is not specified in filter list, it technically is not applied right - i.e. the IR attribute is not generated right? 

I'm not sure if I am misunderstanding the filter string here

Reply:
The intention here is to try and keep the frontend as oblivious to the actual values of the attribute as possible, only creating the name-value pairs it when strictly necessary. Your understanding of the filter is right, but with or without the filter, we know that having a set of name-values pairs on `add_ir_attributes_function` means that we either have `sycl-single-task` alone (implicitly added by the runtime) or some values added as a result of users specifying properties on the kernel. Even if the values of these properties are being filtered out, the was still trying to use a combination of compile-time properties and attributes with compile-time properties alternatives and should be warned as such.

> Does `getAttributeNameValuePairs` return only those in the filter list?

`getAttributeNameValuePairs` returns all name-value pairs. `getFilteredAttributeNameValuePairs`, which I originally used here by accident, returns only the name-value pairs that were allowed through the filter.

> Also if there is a filter list and sycl-single-task is specified in add_ir_attributes_function but is not specified in filter list, it technically is not applied right - i.e. the IR attribute is not generated right?

That is right, but since we are just ignoring it here if it is alone, it does not have much of an effect on this check whether or not it is included. Also, since it is added by the device headers directly in the attribute, it is unlikely to ever not be in the filter if it is in the attribute.

Reply:
IMO you should be using "`getFilteredAttributeNameValuePairs`" since only those name-value pairs can actually result in a conflict. If none of the name-value pairs are applicable on kernel and are instead being applied to say `add_ir_annotations_member`, why should there be a conflict?

Reply:
> Currently we only check to see if we have user-specified compile-time properties and some attributes that could conflict with them. 

Right but the issue here is that we use the user specified properties as a superset for various attributes implementing compile time properties feature.

We're essentially discussing what `could` means here. The way I see it you have 2 lists available to you at the time of this check where you're handling conflicts for only `[[__sycl_detail__::add_ir_attributes_function()]]` - a Subset list (filtered list) and a Superset list

IIUC, as per compile time property design, user specified function attributes can only conflict with the subset list. Of course this means there is a conflict with superset as well but couldn't there be cases where the superset does not contain any properties meant for ` [[__sycl_detail__::add_ir_attributes_function()]]` .  

>>   though at that point we might as well start checking that there are actual conflicts.
I don't think we need to do this.  This would require additional parsing.  I'm asking why we cannot use the information we already have to make the warning fire less.